src/libraries/System.Private.Xml/src/System/Xml/BinaryXml/XmlBinaryReader.cs

```diff
@@ -3523,63 +3512,55 @@ private XmlNodeType ScanOverAnyValue(BinXmlToken token, bool attr, bool checkCha
         }
 
-        private unsafe XmlNodeType CheckText(bool attr)
+        private XmlNodeType CheckText(bool attr)
         {
             Debug.Assert(_checkCharacters, "this.checkCharacters");
-            // assert that size is an even number
-            Debug.Assert(0 == ((_pos - _tokDataPos) & 1), "Data size should not be odd");
             // grab local copy (perf)
 
-            fixed (byte* pb = _data)
-            {
-                int end = _pos;
-                int pos = _tokDataPos;
+            ReadOnlySpan<byte> data = _data.AsSpan(_tokDataPos, _end - _tokDataPos);
+            Debug.Assert(data.Length % 2 == 0, "Data size should not be odd");
 
-                if (!attr)
+            if (!attr)
+            {
+                // scan if this is whitespace
+                while (true)
                 {
-                    // scan if this is whitespace
-                    while (true)
-                    {
-                        int posNext = pos + 2;
-                        if (posNext > end)
-                            return _xmlspacePreserve ? XmlNodeType.SignificantWhitespace : XmlNodeType.Whitespace;
-                        if (pb[pos + 1] != 0 || !XmlCharType.IsWhiteSpace((char)pb[pos]))
-                            break;
-                        pos = posNext;
-                    }
+                    if (!BinaryPrimitives.TryReadUInt16LittleEndian(data, out ushort value))
+                        return _xmlspacePreserve ? XmlNodeType.SignificantWhitespace : XmlNodeType.Whitespace;
+                    if (value > byte.MaxValue || !XmlCharType.IsWhiteSpace((char)value))
+                        break;
+                    data = data.Slice(2); // we consumed one ANSI whitespace char
                 }
+            }
 
+            while (true)
+            {
+                char ch;
                 while (true)
                 {
-                    char ch;
-                    while (true)
-                    {
-                        int posNext = pos + 2;
-                        if (posNext > end)
-                            return XmlNodeType.Text;
-                        ch = (char)(pb[pos] | ((int)(pb[pos + 1]) << 8));
-                        if (!XmlCharType.IsCharData(ch))
-                            break;
-                        pos = posNext;
-                    }
+                    if (!BinaryPrimitives.TryReadUInt16LittleEndian(data, out ushort value))
+                        return XmlNodeType.Text;
+                    data = data.Slice(2); // we consumed one char (possibly a high surrogate)
+                    ch = (char)value;
+                    if (!XmlCharType.IsCharData(ch))
+                        break;
+                }
 
-                    if (!XmlCharType.IsHighSurrogate(ch))
+                if (!XmlCharType.IsHighSurrogate(ch))
+                {
+                    throw XmlConvert.CreateInvalidCharException(ch, '\0', ExceptionType.XmlException);
```

Comment:
seems like previously we explicitly did not move forward on exception but now we do, is there any chance this could cause any issues?

Reply:
I don't think I changed the logical flow of this method. What am I not seeing?

Reply:
sorry, my bad